### PR TITLE
SONARJAVA-1444 Refactor children Iterable, FirstSyntaxTokenFinder and LastSyntaxTokenFinder

### PR DESCRIPTION
**Pull request by mpaladin**

### CODE REVIEW

File: java-squid/src/main/java/org/sonar/java/syntaxtoken/LastSyntaxTokenFinder.java [hunk 1]
```diff
@@ -20,449 +20,62 @@
...
-   * <ul>
-   *   <li>Empty list of modifiers ({@link org.sonar.plugins.java.api.tree.ModifiersTree})</li>
-   *   <li>Any tree of Kind "OTHER" ({@link org.sonar.plugins.java.api.tree.Tree.Kind.OTHER})</li>
-   * </ul>
-   */
-  @Nullable
-  public static SyntaxToken lastSyntaxToken(Tree tree) {
-    if (tree.is(Tree.Kind.TOKEN)) {
+  @CheckForNull
+  public static SyntaxToken lastSyntaxToken(@Nullable Tree tree) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/syntaxtoken/LastSyntaxTokenFinder.java]: MAJOR Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1142)
MAJOR The Cyclomatic Complexity of this method "lastSyntaxToken" is 12 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-squid/src/main/java/org/sonar/java/syntaxtoken/LastSyntaxTokenFinder.java [hunk 2]
```diff
@@ -20,449 +20,62 @@
...
+    if (tree == null) {
+      return null;
+    } else if (tree.is(Tree.Kind.TOKEN)) {
       return (SyntaxToken) tree;
+    } else if (tree.is(Tree.Kind.INFERED_TYPE)) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/syntaxtoken/LastSyntaxTokenFinder.java]: MAJOR This branch's code block is the same as the block for the branch on line 40. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1871)

File: java-squid/src/main/java/org/sonar/java/syntaxtoken/FirstSyntaxTokenFinder.java [hunk 3]
```diff
@@ -19,450 +19,38 @@
...
-   * <ul>
-   *   <li>Empty list of modifiers ({@link org.sonar.plugins.java.api.tree.ModifiersTree})</li>
-   *   <li>Any tree of Kind "OTHER" ({@link org.sonar.plugins.java.api.tree.Tree.Kind.OTHER})</li>
-   * </ul>
-   */
-  @Nullable
-  public static SyntaxToken firstSyntaxToken(Tree tree) {
-    if (tree.is(Tree.Kind.TOKEN)) {
+  @CheckForNull
+  public static SyntaxToken firstSyntaxToken(@Nullable Tree tree) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/syntaxtoken/FirstSyntaxTokenFinder.java]: MAJOR Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1142)
MAJOR The Cyclomatic Complexity of this method "firstSyntaxToken" is 13 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-squid/src/main/java/org/sonar/java/syntaxtoken/FirstSyntaxTokenFinder.java [hunk 4]
```diff
@@ -19,450 +19,38 @@
...
-      firstSyntaxToken = firstModifierToken;
-    } else {
-      TypeParameters typeParameters = tree.typeParameters();
-      if (!typeParameters.isEmpty()) {
-        firstSyntaxToken = typeParameters.openBracketToken();
-      } else if (tree.returnType() != null) {
-        scan(tree.returnType());
-      } else {
-        firstSyntaxToken = tree.simpleName().identifierToken();
+    } else if (tree.is(Tree.Kind.INFERED_TYPE)) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/syntaxtoken/FirstSyntaxTokenFinder.java]: MAJOR This branch's code block is the same as the block for the branch on line 36. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1871)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues:
* BLOCKER 1 blocker
* MAJOR 7 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* BLOCKER [NullPointerException might be thrown as 'FirstSyntaxTokenFinder' is nullable here](https://github.com/SonarSource/sonar-java/blob/94a98a920114824adbcce8567fb4eca4a5c3513f/java-checks/src/main/java/org/sonar/java/checks/HiddenFieldCheck.java#L119) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2259)
* MAJOR [Protected method '<init>(...)' is never used.](https://github.com/SonarSource/sonar-java/blob/94a98a920114824adbcce8567fb4eca4a5c3513f/java-squid/src/main/java/org/sonar/java/model/InternalSyntaxToken.java#L40) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUnusedProtectedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues:
* BLOCKER 1 blocker
* MAJOR 7 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* BLOCKER [NullPointerException might be thrown as 'FirstSyntaxTokenFinder' is nullable here](https://github.com/SonarSource/sonar-java/blob/94a98a920114824adbcce8567fb4eca4a5c3513f/java-checks/src/main/java/org/sonar/java/checks/HiddenFieldCheck.java#L119) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2259)
* MAJOR [Protected method '<init>(...)' is never used.](https://github.com/SonarSource/sonar-java/blob/94a98a920114824adbcce8567fb4eca4a5c3513f/java-squid/src/main/java/org/sonar/java/model/InternalSyntaxToken.java#L40) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUnusedProtectedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* BLOCKER 1 blocker
* MAJOR 3 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* BLOCKER [NullPointerException might be thrown as 'FirstSyntaxTokenFinder' is nullable here](https://github.com/SonarSource/sonar-java/blob/2df52bb34a01db77b25d22b2ce6b5d47ae6aea4a/java-checks/src/main/java/org/sonar/java/checks/HiddenFieldCheck.java#L119) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2259)
* MAJOR [Protected method '<init>(...)' is never used.](https://github.com/SonarSource/sonar-java/blob/2df52bb34a01db77b25d22b2ce6b5d47ae6aea4a/java-squid/src/main/java/org/sonar/java/model/InternalSyntaxToken.java#L40) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUnusedProtectedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* BLOCKER 1 blocker
* MAJOR 3 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* BLOCKER [NullPointerException might be thrown as 'FirstSyntaxTokenFinder' is nullable here](https://github.com/SonarSource/sonar-java/blob/2df52bb34a01db77b25d22b2ce6b5d47ae6aea4a/java-checks/src/main/java/org/sonar/java/checks/HiddenFieldCheck.java#L119) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2259)
* MAJOR [Protected method '<init>(...)' is never used.](https://github.com/SonarSource/sonar-java/blob/2df52bb34a01db77b25d22b2ce6b5d47ae6aea4a/java-squid/src/main/java/org/sonar/java/model/InternalSyntaxToken.java#L40) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUnusedProtectedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* BLOCKER 1 blocker
* MAJOR 3 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* BLOCKER [NullPointerException might be thrown as 'FirstSyntaxTokenFinder' is nullable here](https://github.com/SonarSource/sonar-java/blob/f1c66db5e0f57e28e51e049c4c929b2343366733/java-checks/src/main/java/org/sonar/java/checks/HiddenFieldCheck.java#L119) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2259)
* MAJOR [Protected method '<init>(...)' is never used.](https://github.com/SonarSource/sonar-java/blob/f1c66db5e0f57e28e51e049c4c929b2343366733/java-squid/src/main/java/org/sonar/java/model/InternalSyntaxToken.java#L40) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUnusedProtectedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* BLOCKER 1 blocker
* MAJOR 3 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* BLOCKER [NullPointerException might be thrown as 'FirstSyntaxTokenFinder' is nullable here](https://github.com/SonarSource/sonar-java/blob/f1c66db5e0f57e28e51e049c4c929b2343366733/java-checks/src/main/java/org/sonar/java/checks/HiddenFieldCheck.java#L119) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2259)
* MAJOR [Protected method '<init>(...)' is never used.](https://github.com/SonarSource/sonar-java/blob/f1c66db5e0f57e28e51e049c4c929b2343366733/java-squid/src/main/java/org/sonar/java/model/InternalSyntaxToken.java#L40) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUnusedProtectedMethod)

**Comment by benzonico** [comment]
I have only one question when looking at this PR : why ?

**Comment by mpaladin** [comment]
@benzonico for the record: it removes the complexity of `FirstSyntaxTokenFinder` and `LastSyntaxTokenFinder` which were huge `BaseTreeVisitor`s classes prone to errors, we already have such logic in `childrenIterator` (now `childrenIterable` for flexibility) and its somehow pointless to duplicate it.

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* BLOCKER 1 blocker
* MAJOR 3 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* BLOCKER [NullPointerException might be thrown as 'FirstSyntaxTokenFinder' is nullable here](https://github.com/SonarSource/sonar-java/blob/2e9a41f37018ff74676b5dfade5f0bd98d05d469/java-checks/src/main/java/org/sonar/java/checks/HiddenFieldCheck.java#L119) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2259)
* MAJOR [Protected method '<init>(...)' is never used.](https://github.com/SonarSource/sonar-java/blob/2e9a41f37018ff74676b5dfade5f0bd98d05d469/java-squid/src/main/java/org/sonar/java/model/InternalSyntaxToken.java#L40) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUnusedProtectedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* BLOCKER 1 blocker
* MAJOR 3 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* BLOCKER [NullPointerException might be thrown as 'FirstSyntaxTokenFinder' is nullable here](https://github.com/SonarSource/sonar-java/blob/2e9a41f37018ff74676b5dfade5f0bd98d05d469/java-checks/src/main/java/org/sonar/java/checks/HiddenFieldCheck.java#L119) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2259)
* MAJOR [Protected method '<init>(...)' is never used.](https://github.com/SonarSource/sonar-java/blob/2e9a41f37018ff74676b5dfade5f0bd98d05d469/java-squid/src/main/java/org/sonar/java/model/InternalSyntaxToken.java#L40) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUnusedProtectedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Protected method '<init>(...)' is never used.](https://github.com/SonarSource/sonar-java/blob/229c51d7d9fc65d095f1865c7a64986a11ea435b/java-squid/src/main/java/org/sonar/java/model/InternalSyntaxToken.java#L40) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUnusedProtectedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Protected method '<init>(...)' is never used.](https://github.com/SonarSource/sonar-java/blob/229c51d7d9fc65d095f1865c7a64986a11ea435b/java-squid/src/main/java/org/sonar/java/model/InternalSyntaxToken.java#L40) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUnusedProtectedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Protected method '<init>(...)' is never used.](https://github.com/SonarSource/sonar-java/blob/c1fd2054d09d4141546d8ac9774b62c19be052d7/java-squid/src/main/java/org/sonar/java/model/InternalSyntaxToken.java#L40) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUnusedProtectedMethod)

**Pull request merged**